### ISSUE TITLE
`queue.max_bytes >= queue.page_capacity` check is not done in multipipelines

### ISSUE THREAD

**Issue by andsel** [elastic/logstash#12213]
When Logstash starts it uses https://github.com/elastic/logstash/blob/e9c9865f4066b54048f8d708612a72d25e2fe5d9/logstash-core/lib/logstash/bootstrap_check/persisted_queue_config.rb#L23 to check that `queue.max_size >= queue.page_capacity` configured in `config/logstash.yml` however, if we configure 2 pipelines in  `config/pipelines.yml` each with `queue.type: persisted` and `queue.max_size: 0` Logstash starts cleanly

**verify Logstash does the check**

- edit `config/logstash.yml` setting:
 ```
queue.type: persisted
queue.max_bytes: 0
```
- start logstash with `bin/logstash -e "input{stdin{}} output{stdout}"`
- verify Logstash refuse to start with error `ERROR: Invalid configuration, 'queue.page_capacity' must be less than or equal to 'queue.max_bytes'`

**verify Logstash doesn't  check queue size for multiple pipeline**
- edit 2 files, es `/tmp/pipeline1.conf` and `/tmp/pipeline2.conf` with some simple pipeline config like `input{stdin{}} output{stdout}`
- edit `config/pipelines.yml` to have the content:
```
 - pipeline.id: pipeline1
   queue.type: persisted
   queue.max_bytes: 0
   path.config: "/tmp/pipeline1.conf"
 - pipeline.id: pipeline2
   queue.type: persisted
   queue.max_bytes: 0
   path.config: "/tmp/pipeline2.conf"  
```
- verify Logstash doesn't provide any warning/error and pipelines are  active and with 0 as max_bytes (`curl -XGET 'localhost:9600/_node/stats/pipelines?pretty' | jq .pipelines.pipeline1.queue.max_queue_size_in_bytes` and `curl -XGET 'localhost:9600/_node/stats/pipelines?pretty' | jq .pipelines.pipeline2.queue.max_queue_size_in_bytes`)

The expected behavior is that a configuration like that should prevent the pipeline to start.

**Comment by Crazybus** [comment]
Logstash user checking in as someone that thought this feature was working as intended. `queue.max_events: 0` being documented as the way to set `queue.max_events` to unlimited seemed like a logical configuration to use to also set `queue.max_bytes` to unlimited. In my testing it is working as expected and Logstash persistent queues function correctly and fails in an expected way when the disk is full. I would expect that there are other users out there also happily using this configuration and removing it could be seen as a breaking change.

Some questions from a user point of view:

1. Will this bug fix be treated as a breaking change?
2. Will there be an alternative configuration option provided to set the `queue.max_bytes` to unlimited?
3. When using a queue limit is there a way for the monitoring endpoint used by health checks to detect that back pressure has started? (more info about this below)

---

To give a bit of background for why `queue.max_bytes: 0` and having Logstash crash when disk space runs out is desirable:

Our situation looks like the following:

* All events go to a single Logstash endpoint
* Logstash routes all events to a single output used for alerting via a dedicated pipeline and persistent queue
* Logstash routes each event to a specific Elasticsearch cluster. Each of these outputs has their own dedicated pipeline and persistent queue to isolate output failures

Each output has a dedicated pipeline and persistent queue to limit the blast radius with how [how persistent queues handle back pressure](https://www.elastic.co/guide/en/logstash/current/persistent-queues.html#backpressure-persistent-queue). Where a single queue filling up means that all inputs and outputs fail. Using a single pipeline with a single dedicated persistent queue would mean that a single output would stop all outputs from working. Meaning a single output outage would cause all events to stop flowing, and for our queue to fill up at the rate of all of our outputs combined. As this would interrupt our alerting workflow back pressure is seen as a failure mode for us. The majority of our infrastructure runs on immutable machines too which is another reason for why it is preferable for us to queue events persistently in Logstash.

To prevent this we need to have dedicated pipelines and queues to isolate Logstash from partial output failures. This setup is created to make sure that we have as much time as possible to respond to and fix an output failure before it can impact other outputs. We currently aim to make sure our total queue size combined with our current event rate will give us enough time to last 3 days (long weekend) without manual intervention. E.g. A single Elasticsearch cluster being out means our Alerting will still keep working for 3 days.

With `queue.max_bytes: 0` we are able to achieve an ideal system that makes optimum usage of our available resources to increase the amount of time before total failure. And one that gracefully degrades service in a way that modern infrastructure automation expects. Here is how a single output failure looks with this setup. To make numbers easy we are assuming that we have 3 Logstash nodes, 100gb available for queue space and that we have 10 outputs. We will also assume that each output produces 10gb of logs per hour per node. And that our infrastructure will remove a Logstash node from the loadbalancer if the health check is failing, or the process is not running. 

Happy flow:

* An output starts failing and events start building up in the queue
* Queue growth is detected and an engineer receives an alert
* Looking at the graphs in the alert the engineer can see that the queue will fill up in around 10 hours
* Engineer fixes the output issue within 10 hours and the queue clears up

Sad flow:

* Something goes wrong and after 10 hours the queue does actually fill up the disk despite having significant warning time
* Logstash crashes and events are routed to other still healthy nodes. At this point in time we have lost a node, some events for the broken output are stuck in the queue. But we are still getting the latest events flowing through other nodes for our other outputs. We are missing some historical data, but we at least have the latest information to know our current alerting situation.
* One by one each node exhausts the 100gb disk queue and events stop flowing. We now have no alerting or events available in Elasticsearch for troubleshooting.
* Engineer restores the output, and restores service by adding more nodes and/or increasing the disk space.

In a system where the available disk space needs to be divided between all queues (e.g. `queue.max_bytes: 10gb`)

Happy Flow:

* An output starts failing and events start building up in the queue
* Queue growth is detected and an engineer receives an alert
* Looking at the graphs in the alert the engineer can see that the queue will fill up in only 1 hour
* Engineer fixes the output issue within 1 hour and the queue clears up

Sad flow:

* The engineer can't resolve the issue within 1 hour
* After 1 hour all inputs and outputs stop working on one node
* Although the node is not processing events. It is still running and "green" according to the monitoring API used for the health check. It remains in the loadbalancer and puts back pressure on inputs. At this point we are now missing 1/3 of events for all outputs. Meaning we are missing events for our alerting and dashboards even though only one node has failed. (Note: there may be a way to detect this state, however it is not currently caught by the [default health check used for by the Logstash Helm Charts](https://github.com/elastic/helm-charts/blob/master/logstash/values.yaml#L208-L216).
* Shortly after the other nodes also exhaust their queue and we have no alerting or events.

**Comment by andsel** [comment]
> 1. Will this bug fix be treated as a breaking change?
I don't think this is a breaking change, because it wasn't a feature, but a worked by side effect

> 2. Will there be an alternative configuration option provided to set the queue.max_bytes to unlimited?
we could discuss about this, defining which is the expect behavior.

> 3. When using a queue limit is there a way for the monitoring endpoint used by health checks to detect that back pressure has started? (more info about this below)

yes, the metrics related to queue that Logstash pushes through the monitoring has `max_queue_size_in_bytes` and `queue_size_in_bytes` for each persistent queue of each pipeline, so when `queue_size_in_bytes == max_queue_size_in_bytes` that queue is exerting backpressure on his inputs.

----

Some notes. 
- let the process crashing is not good point, the problem is that a missed flush of data to disk due to IOError could corrupt a queue page, and also it could be that queue checkpoint is not aligned to the queue content, making the subsequent restart of the node a little painfull.
- instead of use `queue.max_bytes: 0` use the size of the disk you have, for example if want to use all the 100Gb of the disk set `queue.max_bytes` to 100Gb minus a little bit of space, Logstash PQ (aka Persistent Queue) has a little file named checkpoint that has to be saved.
- if you have 1 of the 3 nodes that's exerting BP (Back Pressure) you are loosing 1/3 of the events till that 1/3 fills up also the front pipeline, at that point all you processing is stopped. 

A crash could always happen, but forcing a node to crash and  using this event as mechanism to remove a node from a pool seem like an anti-pattern, probably is best that that node signal his "yellow state", still alive but not operating in bad shape  (and potentially could be on track). I know that from the supervisor perspective the result is the same, but it PQ crash badly it's more difficult to recover it. WDYT?

**Comment by Crazybus** [comment]
> 3. When using a queue limit is there a way for the monitoring endpoint used by health checks to detect that back pressure has started? (more info about this below)
>> yes, the metrics related to queue that Logstash pushes through the monitoring has max_queue_size_in_bytes and queue_size_in_bytes for each persistent queue of each pipeline, so when queue_size_in_bytes == max_queue_size_in_bytes that queue is exerting backpressure on his inputs.

Could you give an example of how to do this (especially with multi pipelines)? In an ideal world I would expect for the status in `curl -XGET 'localhost:9600` to not be green in the event that Logstash is not capable of processing events because the queue is full. From a user point of view I'm looking for an automated way to detect this failure scenario so that this Logstash node can be removed from the loadbalancer. That way a single node failure will not result in event lost.

> instead of use queue.max_bytes: 0 use the size of the disk you have, for example if want to use all the 100Gb of the disk set queue.max_bytes to 100Gb minus a little bit of space, Logstash PQ (aka Persistent Queue) has a little file named checkpoint that has to be saved.

Thank you for confirming that the `queue.max_bytes` setting in logstash.yml is supposed to be a global limit for all queues. When I first tested this out (what I tried before setting it to 0 to make it unlimited) it didn't work. My initial testing made me think that this limit was just the default for each pipeline. And that setting it to 2gb would require 3x2gb of total space (6gb in total). I did some more digging to try to understand why it isn't working as expected for me.

This is my test setup:

* Logstash 7.9.1 running in Docker
* Dedicated queue volume of 2gb `mount -t tmpfs -o size=2g tmpfs $(realpath queue)`
* 2 persistent queue pipelines configured
* `queue.max_bytes: 1500mb` in logstash.yml
* `queue.max_events: 0` logstash.yml
* An input pipeline that just sends heartbeats as fast as possible to fill the queue
* A broken output pointing to Elasticsearch to force the queue to be used up
* A working output that just logs to a file so I can detect when the queue has been blocked.

When I first started up Logstash it worked this time. I was really confused since in my previous testing I wasn't able to set a `queue.max_bytes` that was larger than total disk space - (max_bytes * persistent queues).

Once the queue was filled up a bit (~600mb) I restarted Logstash and I got this error.

```
[ERROR] 2020-09-07 12:19:34.542 [Converge PipelineAction::Create<riemann>] agent - Failed to execute action {:action=>LogStash::PipelineAction::Create/pipeline_id:riemann, :exception=>"Java::JavaLang::IllegalStateException", :message=>"java.io.IOException: Unable to allocate 1538419593 more bytes for persisted queue on top of its current usage of 34444407 bytes", :backtrace=>["org.logstash.execution.AbstractPipelineExt.openQueue(AbstractPipelineExt.java:204)", "org.logstash
```

So it seems like Logstash is failing to start up because it actually wants to have `queue.max_bytes` + `queue.currently_used_bytes` available at start up. This check was added in https://github.com/elastic/logstash/pull/10052 but it doesn't seem to be calculating the available space correctly.

In my previous testing I already would have had a fullish queue which explains why I couldn't get Logstash to start with global `queue.max_bytes` limits being set despite having enough total disk space. Which is why I made the assumption that this max_bytes was a default setting for each pipeline instead of a global limit.

Is this just a bug in how Logstash is calculating available space? In my testing (and in production) I'm using a dedicated volume for the queue and nothing else is writing to it.

> A crash could always happen, but forcing a node to crash and using this event as mechanism to remove a node from a pool seem like an anti-pattern, probably is best that that node signal his "yellow state", still alive but not operating in bad shape (and potentially could be on track). I know that from the supervisor perspective the result is the same, but it PQ crash badly it's more difficult to recover it. WDYT?

100% agree that crashing is not the most graceful solution to handling a node failure. However at this stage it is still more preferable than having 1/3 of events being lost due to back pressure. In a highly available system a single node failure should only result in capacity loss, and not in event loss. Being able to use all of the disk space is a way for us to delay back pressure from ever kicking in. At the moment it seems like we are still looking for a way to:

1. Use all of the available disk space for queueing events (to prolong BP failure scenario). If my understanding is correct this should be possible with setting it in logstash.yml but there appears to be a bug with how available queue space is being calculated.
2. A way for Logstash to signal that it is not healthy and should be removed from the loadbalancer. Ideally this would be reported by the healthcheck endpoint from Logstash which currently reports a green status even though it is unable to process events. Another alternative would be for Logstash to stop listening on its inputs completely. E.g. the http input port stops listening, right now it still responds with `ok` which signals to the loadbalancer that it is still alive and healthy.

**Comment by andsel** [comment]
> Could you give an example of how to do this (especially with multi pipelines)?
you can use the http api an jq, but you still need some code. 
```sh
```

**Comment by Crazybus** [comment]
> you can use the http api an jq, but you still need some code.

Thanks! The official Logstash images don't include jq and this example only works for a single pipeline named test. The official images include Python2 so I made a quick script that will work for a setup with multiple piplines:

```
#!/usr/bin/env python2

import json
import urllib2

r = json.loads(urllib2.urlopen("http://127.0.0.1:9600/_node/stats").read())

# Percentage full for a queue being considering it not healthy
threshold = 95

exit_code = 0

for pipeline, data in r["pipelines"].iteritems():
    q = data["queue"]

    # Skip in memory queues for checking the byte limits
    if q["type"] == "memory":
        continue

    queue_size = float(q["capacity"]["queue_size_in_bytes"])
    max_queue_size = float(q["capacity"]["max_queue_size_in_bytes"])
    percent = int(queue_size / max_queue_size * 100)

    print("{0}:\t\t {1}%".format(pipeline, percent))

    if percent >= threshold:
        exit_code = 1
        print(
            "ERROR: Pipeline {0} is at {1}% which is over threshold of {2}%".format(
                pipeline, percent, threshold
            )
        )

exit(exit_code)
```

> The setting queue.max_bytes in logstash.yml is intended to be the value is not differently specified in pipelines.yml and not the sum of all pipelines' bytes. So if set it to 2Gb and have 2 pipelines it means at each pipelines could have a queue of 2Gb (totally 4Gb).

This doesn't seem to line up with the advice you gave in your previous comment:

> instead of use queue.max_bytes: 0 use the size of the disk you have, for example if want to use all the 100Gb of the disk set queue.max_bytes to 100Gb minus a little bit of space, Logstash PQ (aka Persistent Queue) has a little file named checkpoint that has to be saved.

Which was recommending that I set `queue.max_bytes` to be something like 99gb to use all of my 100gb disk. Looking at the monitoring data field `max_queue_size_in_bytes` it seems like I'm telling Logstash it is able to use 198gb out of the available 100gb disk space. The healthcheck you are recommending would mean that if both queues were filling up an equal rate that the disk space would run out at around 50% which wouldn't be caught because the health check is only firing at 90%.

This brings me back to why `queue.max_bytes: 0` is something I would like to see supported and should maybe even be the default. The current implementation of `queue.max_bytes > 0` requires users to artificially limit the available queue capacity. Having to divide up the available disk space amongst all of the pipelines only increases the time to failure when a queue is filling up. What is the advantage of enforcing this limit? As an operator I don't want Logstash to start failing prematurely when there is still plenty of disk space leftover.

From what I understand the only concern with `queue.max_bytes: 0` is around Logstash crashing because it currently doesn't check the available disk space when attempting to do disk writes. Which sounds like a separate issue that should be resolved as well since there are other scenarios where the disk could run out of space such as:

* The queue is being stored on a shared disk
* Some of the disk space has been unexpectedly taken away by an old queue that hasn't been cleaned up. When removing a pipeline configuration from Logstash the old queue is still left on disk.

In the short term this means our health check should just check if there is enough disk space. And in the long term this could be improved in Logstash by:

* Making sure Logstash can handle running out of disk space without crashing
* Invoking backpressure when the disk space where the queue is located is over a certain threshold
* Having a health check that will report an error when backpressure has been invoked

> However the discussion of yellow state if one persistent queue exert backpressure, require some discussion to understand better if it's needed and how should be implemented.

Could you explain what you mean by "if one persistent queue exert backpressure"? With the current design what I am experiencing is that a single queue hitting the back pressure conditions causes all pipelines to fail (including non-persistent in memory queues). From an operator point of view Logstash is unable to do anything in its current state.

**Comment by jsvd** [comment]
I'd like to chime in here a bit with some history of why the knobs and metrics are the way they are, and some thoughts on what they should be in the future.

When we built the PQ back in 5.1 (+3 years ago), Logstash only supported one pipeline and it was very common for it to be deployed alongside other software, as we were "beginning" to push filebeat as the right way to tail a file.

During development we had the discussion on how to set the size limits for the PQ.Setting a limit based on amount of disk free (e.g. it's full if the disk only has 5% free), could lead to Logstash using most of the space that was shared with a user's Tomcat or Apache Server.
So considering most disks nowadays had > 100gb, we capped the PQ itself to be 1gb. This would allow it to absorb small bursts of traffic, and anything more prolonged would be back pressured, thus avoiding logstash having to deal with large backlogs of data while not putting tens or hundreds of GB at risk of data loss since the PQ isn't replicated. 

Now back to the future, Logstash supports multiple pipelines, and it's no longer sharing resources. As you mentioned already the current knobs can create false sense of resiliency, especially for large PQs or a large number of PQs that, in total, can use close to or more than the available disk space.
So at this point it's best to monitor how full the queues and the disk are, and take action on the first to hit a threshold like 90%.

> Could you explain what you mean by "if one persistent queue exert backpressure"? With the current design what I am experiencing is that a single queue hitting the back pressure conditions causes all pipelines to fail (including non-persistent in memory queues). From an operator point of view Logstash is unable to do anything in its current state.

Just to confirm, this happens when pipelines are connected to other pipelines, and not to parallel pipelines whose data flow isn't linked together, right?
Since logstash doesn't allow two outputs in 1 pipeline to move at different speeds, then as long as there's a data flow between pipelines then backpressure will be exerted all the way through them back to the data source.

**Comment by Crazybus** [comment]
> I'd like to chime in here a bit with some history of why the knobs and metrics are the way they are, and some thoughts on what they should be in the future.

Thanks a lot for the fun history lesson! This certainly helps to explain how the limits currently work.

> So at this point it's best to monitor how full the queues and the disk are, and take action on the first to hit a threshold like 90%.

Thanks! This is what we will go with for now until there is something natively built into Logstash to handle it.

> Thinking forward we should contemplate moving to thresholds based on disk free space watermarks, either as a replacement or in complement to the existing knobs.

Sounds perfect :)

> Just to confirm, this happens when pipelines are connected to other pipelines, and not to parallel pipelines whose data flow isn't linked together, right?
> Since logstash doesn't allow two outputs in 1 pipeline to move at different speeds, then as long as there's a data flow between pipelines then backpressure will be exerted all the way through them back to the data source.

Correct, each output is isolated into its own pipeline. Otherwise a single output going down means all outputs in that pipeline immediately stop working.

However it also depends on your definition of linked. I described our setup in more detail in https://github.com/elastic/logstash/issues/12213#issuecomment-686497229. The TL;DR from that comment is:

> * All events go to a single Logstash endpoint
> * Logstash routes all events to a single output used for alerting via a dedicated pipeline and persistent queue
> * Logstash routes each event to a specific Elasticsearch cluster. Each of these outputs has their own dedicated pipeline and persistent queue to isolate output failures

So in this case a single pipeline having its queue fill up will mean that the backpressure applies back to the original input pipeline. So once any output queue has been filled all inputs block. From a systems design point of view this makes perfect sense and is expected. If Logstash can no longer queue events because there is no more space then it should stop working. However we would like to make sure there is a supported way to use all of the available queuing space before hitting this failure scenario. And that once it is hit it can be detected properly via a health check.

**Comment by colinsurprenant** [comment]
Hello, thanks all for the discussion and @Crazybus for raising valid concerns. 
I'd like to take a step back and reframe this around the original issue filed by @andsel and talk about `queue.max_bytes`.

We definitely have a few problems on our hands here. 

- The original intent for `queue.max_bytes: 0` was to allow an unbounded queue size, similar to `queue.max_events: 0`, see `Queue#isFull()` method: https://github.com/elastic/logstash/blob/164a763d7e8c33e1dcee73a909f827ee07c74b12/logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java#L473

  Note that this is not explicitly documented like `queue.max_events: 0` is. 

- A check for available disk space is done at `Queue#open()` and if `queue.max_bytes: 0` it *should* never fail that check; in fact I think there is a bug in that code where if the current PQ size is actually bigger than the queue max_bytes then  the available disk space check will always pass: https://github.com/elastic/logstash/blob/164a763d7e8c33e1dcee73a909f827ee07c74b12/logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java#L826
 
- There is **bootstrap** check on `queue.max_bytes` done prior to launching any pipeline (and thus prior to initializing any Queue) which checks to make sure that the `queue.page_capacity` is smaller than `queue.max_bytes`; the intent here is straightforward but there are 2 problems with this check: **a)** this is not taking into account the meaning of  `queue.max_bytes: 0`  and **b)** this is only taking into account the global `logstash.yml` `queue.max_bytes` setting 
 and totally ignoring the multiple pipelines usage with `pipelines.yml`:
https://github.com/elastic/logstash/blob/164a763d7e8c33e1dcee73a909f827ee07c74b12/logstash-core/lib/logstash/bootstrap_check/persisted_queue_config.rb#L23


My opinion is that we should allow (and properly document) using `queue.max_bytes: 0` to enable and unbounded queue in byte size. This would essentially mean adjusting the bootstrap check accordingly and possibly refactoring the around the `ensureDiskAvailable` method to render this condition explicit. Now we would have to decide if this is a breaking change and when to introduce it. We should also fix the single/multi pipeline inconsistency in the bootstrap check; I am usure on the right fix for this as this bootstrap check is done prior to having knowledge of the multi pipelines etc. 

The second discussion in this thread is around the free disk space check that should potentially be performed at runtime. To me this is a separate discussion and we should open a new issue about that. There is also some history around that, in particular https://github.com/elastic/logstash/issues/6612 - I am personally in favour of introducing some kind of free disk available check, maybe at new page creation, to **try to avoid** (**try** is important here since logstash might not be the only actor in the system) creating a disk full situation which can really wreak havoc a system in many nasty ways.

**Comment by andsel** [comment]
I agree on the fact this thread shows up some points of improvement.
I would focus the attention on the first, the queue full condition, that was the reason why the `max_bytes: 0` to mean unbounded queue, was more a side feature than a planned feature. 
My concerns is that if:
-  we have `queue.max_events: 0` _and_
-  `queue.max_bytes: 0` _and_ 
- the partition containing the queue has no free space

this should be a "queue full condition" but actually  the code doesn't recognize it, so at the next write request, supposing no reads comes in the middle, check if head page has space, suppose the head page is full then it tries to allocate a new one:

https://github.com/elastic/logstash/blob/164a763d7e8c33e1dcee73a909f827ee07c74b12/logstash-core/src/main/java/org/logstash/ackedqueue/io/MmapPageIOV2.java#L189

and this fail with exception due to disk full.

With a bounded queue when the queue reach it's `max_bytes` limit it exert back pressure, simply waiting on not full condition; but in case of unbounded, when it runs out of disk space it doesn't put on wait condition but throws an error the next time it tries to allocate a page.

**Comment by colinsurprenant** [comment]
@andsel 

First lets make sure we share a common vocabulary: there is "queue full" condition and "disk full" condition, and both are different; "queue full" means that the queue has reached a limit either in configured max byte size or max (unread) events count. "disk full"  is self explanatory. There is no relation between a "queue full" condition and a "disk full" condition. 

- `queue.max_bytes: 0` was not a side effect but in fact a planned behaviour, we explicitly  test for this condition in the `isFull()` method **and** the method description describes it as such: https://github.com/elastic/logstash/blob/164a763d7e8c33e1dcee73a909f827ee07c74b12/logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java#L455-L456
  It was just not properly documented and **not** correctly taken into account in the bootstrap checks which were added later.

Now, WRT the free disk space check you mention, as I said, to me this is a separate topic which is not related to the "queue full" condition. There are 2 places where free disk space checks can be made, the first place is at startup to make sure there is enough disk space to allow the queue to grow to max capacity as defined by `queue.max_bytes`. Per my suggestion above, in my opinion we should not do that check if `queue.max_bytes: 0` is set. The second place is at new page creation where a check **could** be made to see if there is enough disk space to allow for the new page size, but this was never implemented and we have never reached consensus WRT to the best way to handle free disk space checks, see #6612 for some history around that. 

Also note that even when using a non-zero value for `queue.max_bytes` this is really not a bullet proof solution against disk full conditions; there are other components in the system outside logstash that could reduce available disk space after the bootstrap check is done and so without doing checks at page creation, we could end up with a disk full condition. I suggest we open a new discussion issue around that topic, we can certainly revisit this.

**Comment by andsel** [comment]
Resuming the discussion
we have a couple of bugs :

- handling the `max_bytes = 0` with the "unlimited space": the check https://github.com/elastic/logstash/blob/164a763d7e8c33e1dcee73a909f827ee07c74b12/logstash-core/lib/logstash/bootstrap_check/persisted_queue_config.rb#L23 done during Logstash bootstrap (before queue is created) doesn't take care of the unlimited condition, raising an error if `max_bytes=0`.
- ensuring there is enough disk free space during queue opening https://github.com/elastic/logstash/blob/164a763d7e8c33e1dcee73a909f827ee07c74b12/logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java#L826  works correctly for the case`max_bytes=0`, it returns always `ŧrue`, **but** has a problem when `max_bytes < current_queue_bytes`. Could this condition happen only when a user shirk the size of a bounded queue? For example it start with 2Gb, the data fills the queue up to 1,5Gb Logstash is shutdown, the user reduce the amount for `max_bytes` and then restart Logstash.


There are some open discussion:
- in case of unbounded queue with `max_bytes=0` which is the condition to exert back pressure? Actually the "disk full" condition raise an exception, while the "queue full" condition should start the writers to experience the back pressure
- in case we have multipipelines and those defines bounded persistent queues, we don't evaluate if the disk has enough free space for the all the PQs. For example, supposing 2 pipelines each with 40Gb max_bytes, persisted on a disk of 70Gb, actually Logstash permit to start since 40Gb < 70Gb, but the check should be 2*40Gb < 70Gb. Today we can go in weird situations like, we permit Logstash to start, but once the PQs are filled a bit and Logstash is restarted, the free space check fails and the process can't be anymore started, becuase of a condition like "not enough space".
